### ISSUE TITLE
Material copy from a webgl to a webgpu node material break 

### ISSUE THREAD

**Issue by Samsy** [mrdoob/three.js#29973]
### Description

Copying a model to get its properties assigned from a meshmaterial to a nodematerial breaks

### Reproduction steps

1. create a MeshBasicMaterial 
2.create a MeshBasicNodeMaterial
3.copy the meshbasicmaterial into the node one

### Code

`import * as THREE from 'three';
import { OrbitControls } from 'three/addons/controls/OrbitControls.js';

let mesh, renderer, scene, camera, controls;

init();

function init() {

    // renderer
    renderer = new THREE.WebGPURenderer();
    renderer.setSize( window.innerWidth, window.innerHeight );
    renderer.setPixelRatio( window.devicePixelRatio );
    renderer.setAnimationLoop( animate );
    document.body.appendChild( renderer.domElement );

    // scene
    scene = new THREE.Scene();
    scene.background = new THREE.Color( 0x000000 );
    
    // camera
    camera = new THREE.PerspectiveCamera( 40, window.innerWidth / window.innerHeight, 1, 10000 );
    camera.position.set( 20, 20, 20 );

    // controls
    controls = new OrbitControls( camera, renderer.domElement );
    
    // ambient
    scene.add( new THREE.AmbientLight( 0x222222 ) );
    
    // light
    const light = new THREE.DirectionalLight( 0xffffff, 1 );
    light.position.set( 20,20, 0 );
    scene.add( light );
    
    // axes
    scene.add( new THREE.AxesHelper( 20 ) );

    // geometry
    const geometry = new THREE.SphereGeometry( 5, 12, 8 );
    
    // material
    const material = new THREE.MeshBasicMaterial( {
        color: 0x00ffff, 
        flatShading: true,
        transparent: true,
        opacity: 0.7,
    } );
    
    const webGPUMat = new THREE.MeshBasicNodeMaterial()
    
    // here 
    webGPUMat.copy( material )
    
    // mesh
    mesh = new THREE.Mesh( geometry, webGPUMat );
    scene.add( mesh );
    
}

function animate() {

    renderer.render( scene, camera );

}
`

### Live example

[Live fiddle test ](https://jsfiddle.net/njf1mae9/8/)

Line 53 

### Screenshots

<img width="425" alt="Screenshot 2024-11-26 at 11 28 01" src="https://github.com/user-attachments/assets/2efff5f5-5e6a-46bf-90ee-4d4fabbbc142">

### Version

r170

### Device

Desktop

### Browser

Chrome

### OS

MacOS

**Comment by Samsy** [comment]
Ok I understand, 
I'll just override the copy with all the test needed, 
Thought, the copy would be fixed by testing if any property exists in both source and target, it would enabled the copy from any material to any material ?

Apart from this, the problem in this particular case would be that it breaks the app entirely when calling copy from a different material

**Comment by Samsy** [comment]
This may cause a bit of trouble for a lot of usecase

For example, if you load a GLTF using the GLTFLoader, it does not returns a node material 

Thought, if you want to override the material, but copy the property of the original, you don't know what comes from the source material, and which property are set ( aomap, lightmap, normalmap, etc.. ) also if you need to change nature type of the material, same trouble, it ends up being a one for one case of testing properties

**Comment by Mugen87** [comment]
> For example, if you load a GLTF using the GLTFLoader, it does not returns a node material

In general, `WebGPURenderer` can display assets return from `GLTFLoader` without further adjustments.

However, I understand the use case where developers want to load assets via `GLTFLoader` but now apply material modifications. The problem is that a simple copy is not appropriate to transform a normal material to a node material. E.g. in `MeshBasicMaterial` you have `color` and `map` but with `MeshBasicNodeMaterial` you just have a `colorNode` property. For a real node representation you must setup the material differently.

**Comment by Makio64** [comment]
@Mugen87 is the port of GltfLoader to return node based material is on the timeline ?
I encountered similar issue on a project and did a manual translation but it would be nice if it was support out of the box when we're using webgpu three.

**Comment by Mugen87** [comment]
I'm not sure we have discussed this aspect yet. AFAIK, it is not planned so far that loaders return node materials. 

Maybe we can provide some tooling that helps in the conversion process. E.g. a utility method that converts a material type to its node material pendant.

**Comment by Makio64** [comment]
> I'm not sure we have discussed this aspect yet. AFAIK, it is not planned so far that loaders return node materials.
> 
> Maybe we can provide some tooling that helps with the conversion process. E.g. a utility method that converts a material type to its node material pendant.

But then it means we embed all the old system (chunks shaderlib, etc.. ) when we're using a loader ?  😱

**Comment by Mugen87** [comment]
No, I don't think so. The shader chunk system is related to the renderer.

**Comment by DennisSmolek** [comment]
> One reason this might be advantageous compared to a material conversion utility would be to support proposed MaterialX shaders / procedural materials contained directly in glTF.

From my understanding this isn't stable yet.

Neither is how systems interface with it; will the nodes in the gltf be output as attributes of a regular material with the nodes acting as backing for how those attributes were derived? This would mean the system can read the output material as the simpler standard gltf BSDF. 

Or will the system be required to read the MX nodes and generate the material itself? This would mean either using the MX loader code or (like three currently does) use it's own spec conforming loader. Problem there is, the GLTF MX stuff (last time I checked) was going to convert the materialx XML to JSON https://github.com/KhronosGroup/glTF-MaterialX-Converter
Meaning the Matx Loader will have to understand both.

There's a lot to that and getting that to work.

The next thing is the vast majority of systems wont be outputting in MaterialX. Blender for example will still just be the principled BSDF and regular style materials. 

---

There will be some need to convert existing threeJS materials to their node versions and it would be helpful if the library supports it, otherwise users will just make this functionality themselves. 

In regards to the loader (#30114) I think that outputting directly to a nodeMaterial there makes sense too, in a system where you expect all the materials to be nodeMaterials etc it would be a pain if the loader doesnt. You'd essentially traverse immediately and convert.
